public/js/utils/source-map.js

```diff
@@ -1,160 +1,206 @@
-const getOriginalTexts = sourceMapTask("getOriginalTexts");
-const createSourceMap = sourceMapTask("createSourceMap");
-const clearData = sourceMapTask("clearData");
+  if (sourceMap.sourceRoot) {
+    sourceMap.sourceRoot = path.join(base, sourceMap.sourceRoot);
+  } else {
+    sourceMap.sourceRoot = base;
+  }
 
-function _shouldSourceMap(source) {
-  return isEnabled("sourceMaps") && source.sourceMapURL;
+  return sourceMap;
 }
 
-function isMapped(source) {
-  return _shouldSourceMap(source);
+function originalToGeneratedId(originalId) {
+  const match = originalId.match(/(.*)\/originalSource/);
+  return match ? match[1] : null;
 }
 
-function isOriginal(originalSource) {
-  return !!getGeneratedSourceId(originalSource);
+function generatedToOriginalId(generatedId, url) {
+  return generatedId + "/originalSource-" + md5(url);
 }
 
-function isGenerated(source) {
-  return !isOriginal(source);
+function isOriginalId(id) {
+  return id.match(/\/originalSource/);
 }
 
-async function getOriginalSources(state: AppState, source: any) {
-  const originalSourceUrls = await getOriginalSourceUrls(source);
-  return originalSourceUrls.map(url => getSourceByURL(state, url));
+function isGeneratedId(id) {
+  return !isOriginalId(id);
 }
 
-function getGeneratedSource(state: AppState, source: any) {
-  if (isGenerated(source)) {
-    return source;
+function fetchSourceMap(generatedSource) {
+  // If it has already been requested, return the request. An
+  // important behavior here is that if it's in the middle of
+  // requesting it, all subsequent calls will block on the initial
+  // request.
+  const existingRequest = sourceMapRequests.get(generatedSource.id);
+  if (existingRequest) {
+    return existingRequest;
+  } else if (!generatedSource.sourceMapURL || !isEnabled("sourceMaps")) {
+    return Promise.resolve(null);
   }
 
-  const generatedSourceId = getGeneratedSourceId(source);
-  const originalSource = getSource(state, generatedSourceId);
+  async function request() {
+    // Fetch the sourcemap over the network and create it.
+    const sourceMapURL = _resolveSourceMapURL(generatedSource);
+    const fetched = await networkRequest(
+      sourceMapURL, { loadFromCache: false }
+    );
 
-  if (originalSource) {
-    return originalSource.toJS();
+    // Create the source map and fix it up.
+    const map = new SourceMapConsumer(fetched.content);
+    _setSourceMapRoot(map, sourceMapURL, generatedSource);
+    return map;
   }
 
-  return source;
+  // Fire off the request, set it in the cache, and return it.
+  const req = request();
+  sourceMapRequests.set(generatedSource.id, req);
+  return req;
+}
+
+function getSourceMap(generatedSourceId) {
+  return sourceMapRequests.get(generatedSourceId);
 }
 
-async function getGeneratedLocation(state: AppState, location: Location) {
-  const source: any = getSource(state, location.sourceId);
+function installSourceMappings(generatedId, url, code, mappings) {
```

Comment:
Totally. That's the least thought-through part. That name will change. I'm not sure about `create*` because I want to make it clear that it has a side-effect (adds a source map to the system).